crates/sui-core/src/consensus_handler.rs

```diff
@@ -829,6 +858,104 @@ impl ConsensusCommitInfo {
     }
 }
 
+/// Handles certified and rejected transactions output by consensus.
+pub(crate) struct ConsensusTransactionHandler {
+    /// Whether to enable handling certified transactions.
+    enabled: bool,
+    /// Per-epoch store.
+    epoch_store: Arc<AuthorityPerEpochStore>,
+    /// Enqueues transactions to the transaction manager via a separate task.
+    transaction_manager_sender: TransactionManagerSender,
+    /// Metrics for consensus transaction handling.
+    metrics: Arc<AuthorityMetrics>,
+}
+
+impl ConsensusTransactionHandler {
+    pub fn new(
+        epoch_store: Arc<AuthorityPerEpochStore>,
+        transaction_manager_sender: TransactionManagerSender,
+        metrics: Arc<AuthorityMetrics>,
+    ) -> Self {
+        Self {
+            enabled: epoch_store.protocol_config().mysticeti_fastpath(),
+            epoch_store,
+            transaction_manager_sender,
+            metrics,
+        }
+    }
+
+    pub fn enabled(&self) -> bool {
+        self.enabled
+    }
+
+    pub async fn handle_consensus_transactions(&self, parsed_transactions: Vec<ParsedTransaction>) {
+        let mut pending_consensus_transactions = vec![];
+        let executable_transactions: Vec<_> = parsed_transactions
+            .into_iter()
+            .filter_map(|parsed| {
+                // TODO(fastpath): unlock rejected transactions.
+                // TODO(fastpath): maybe avoid parsing blocks twice between commit and transaction handling?
+                if parsed.rejected {
+                    self.metrics
+                        .consensus_transaction_handler_processed
+                        .with_label_values(&["rejected"])
+                        .inc();
+                    return None;
+                }
+                self.metrics
+                    .consensus_transaction_handler_processed
+                    .with_label_values(&["certified"])
+                    .inc();
+                match &parsed.transaction.kind {
+                    ConsensusTransactionKind::UserTransaction(tx) => {
+                        if tx.contains_shared_object() {
```

Comment:
are we sure it's safe to always assume that any tx in consensus without shared objects is a fastpath tx that should be executed here?

this was a little unintuitive to me at first, i wonder if it's worth a more specific name or a comment or something

Reply:
Yes, `contains_shared_object()` is the same check used in `AuthorityState::execute_certificate()`. Definitely we should think of using a separate function for this in future. Added a TODO.